app/assets/javascripts/protocols/steps.js.erb

```diff
@@ -4,637 +4,673 @@
+  function newStepHandler() {
+    $("[data-action='new-step']").off().on('click', function(event) {
+      event.preventDefault();
+      event.stopPropagation();
+      event.stopImmediatePropagation();
+      var $btn = $(this);
+      $btn.off();
+      animateSpinner(null, true);
 
-    var contents = $(this).find('.hot-contents');
-    var data = null;
-    if (contents.attr("value")) {
-      data = JSON.parse(contents.attr("value")).data;
-    }
-    if ($container.is(":visible")) {
-      $(this).css("width", $("#new-step-tables").css("width"));
-    }
+      $.ajax({
+        url: $btn.data('href'),
+        method: 'GET',
+        success: function(data) {
+          var $form = $(data.html);
+          $('#steps').append($form).promise().done(function() {
+            animateSpinner(null, false);
+            // Scroll to bottom
+            $('html, body').animate({
+              scrollTop: $(document).height() - $(window).height()
+            });
+            formCallback($form);
+            formNewAjax($form);
+            applyCancelOnNew();
+            toggleButtons(false);
+            initializeCheckboxSorting();
+
+            $('#step_name').focus();
+            $('#new-step-main-tab a').on('shown.bs.tab', function() {
+              $('#step_name').focus();
+            });
+
+            TinyMCE.refresh();
+          });
 
-    $container.handsontable({
-      data: data,
-      startRows: <%= Constants::HANDSONTABLE_INIT_ROWS_CNT %>,
-      startCols: <%= Constants::HANDSONTABLE_INIT_COLS_CNT %>,
-      minRows: 1,
-      minCols: 1,
-      rowHeaders: true,
-      colHeaders: true,
-      contextMenu: true,
-      formulas: true,
-      preventOverflow: 'horizontal'
+        },
+        error: function() {
+          newStepHandler();
+        }
+      })
     });
+  }
 
-    $container.attr("initialized", true);
-    renderTable($container);
-  });
+  // Needed because server-side validation failure clears locations of
+  // files to be uploaded and checklist's items etc. Also user
+  // experience is improved
+  global.processStep = function processStep(ev, editMode) {
+    ev.stopPropagation();
+
+    var $form = $(ev.target.form);
+    $form.clearFormErrors();
+    $form.removeBlankFileForms();
+
+    var $fileInputs = $form.find("input[type=file]");
+    var filesValid = filesValidator(ev, $fileInputs, FileTypeEnum.FILE);
+    var $checklists = $form.find(".nested_step_checklists");
+    var checklistsValid = checklistsValidator(ev, $checklists, editMode);
+    var $nameInput = $form.find("#step_name");
+    var nameValid = textValidator(ev, $nameInput, 1,
+      <%= Constants::NAME_MAX_LENGTH %>);
+    var $descrTextarea = $form.find("#step_description");
+    var descriptionValid = textValidator(ev, $descrTextarea, 0,
+      <%= Constants::TEXT_MAX_LENGTH %>);
+
+    if (filesValid && checklistsValid && nameValid && descriptionValid) {
+      // Local file uploading
+      animateSpinner();
```

Comment:
I think this is not needed, because `newStepHandler();` already enables and disables spinner by itself.